### PR TITLE
ApplyMeshConfig allow overriding with default value

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/25503

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/config/mesh/mesh.go [hunk 1]
```diff
@@ -122,27 +123,46 @@ func ApplyProxyConfig(yaml string, meshConfig meshconfig.MeshConfig) (*meshconfi
...
+	return string(bytes), nil
+}
+
 // ApplyMeshConfig returns a new MeshConfig decoded from the
 // input YAML with the provided defaults applied to omitted configuration values.
 func ApplyMeshConfig(yaml string, defaultConfig meshconfig.MeshConfig) (*meshconfig.MeshConfig, error) {
-	if err := gogoprotomarshal.ApplyYAML(yaml, &defaultConfig); err != nil {
-		return nil, multierror.Prefix(err, "failed to convert to proto.")
-	}
+	// We want to keep semantics that all fields are overrides, except proxy config is a merge. This allows
```
Comment (howardjohn) [inline pkg/config/mesh/mesh.go:145]: note: I am not changing the semantics here, just documenting them. The only thing I have changed is allowing overrides with default (as defined by go zero values) values

**Comment by nrjpoddar** [comment]
A default "zero" value proto comes back to byte us. Too late for API changes?

**Comment by nrjpoddar** [comment]
Shouldn't the same issue happen for https://github.com/istio/api/blob/master/mesh/v1alpha1/proxy.proto#L336?

**Comment by howardjohn** [comment]
> A default "zero" value proto comes back to bite us. Too late for API change to have a unused zero value?

I agree. We can technically change it but its not wire compatible. We had some people complain about this before - but have also done it before. Technically we don't rely on sending mesh config over the wire

**Comment by nrjpoddar** [comment]
> > A default "zero" value proto comes back to bite us. Too late for API change to have a unused zero value?
> 
> I agree. We can technically change it but its not wire compatible. We had some people complain about this before - but have also done it before. Technically we don't rely on sending mesh config over the wire

I would vote for changing the API. We are making a new minor version, now is the time to address it correctly.

**Comment by howardjohn** [comment]
Neeraj even if we change the API we still need this pr, unless we change every single field to a wrapper field

**Comment by shamsher31** [comment]
> Neeraj even if we change the API we still need this pr, unless we change every single field to a wrapper field

I have tried that already here https://github.com/istio/istio/issues/25503#issuecomment-659940867 

> I have also tried wrapping enum types with &wrappers.UInt32Value{} to see if it persist default values but that is incompatible with ControlPlaneAuthPolicy because it has a custom type int32.


and as @howardjohn said it requires more time and effort.

**Pull request merged**